052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,167 @@
+
+![Proposal](images/052-proposal.png)
+
+Today to enable jmx_exporter, users use:
+```
+metricsConfig:
+  type: jmxPrometheusExporter
+  valueFrom:
+    configMapKeyRef:
+      name: kafka-metrics
+      key: kafka-metrics-config.yml
+```
+
+To enable metric reporters, they will instead use this for example:
+```
+metricsConfig:
+  type: strimziMetricsReporter
+  config:
+    allowList: ".*"
+```
+
+Reporters will expose the following configurations:
+- `prometheus.metrics.reporter.port`: The HTTP port to expose the metrics. Default: `8080`
+- `prometheus.metrics.reporter.allowlist`: A comma separated list of regex patterns to specify the metrics to collect. Default: `.*`. The patterns must match the metric names emitted by the reporter.
+
+Strimzi should set the port to 9404 to make the migration from jmx_exporter easy.
+
+The reporter will also export JVM metrics similar to the ones exported by jmx_exporter. These are provided by the [hotspot exports](io.prometheus.client.hotspot) from the Prometheus Java client.
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it now or later to the reporter.
+
+In Kafka, metric reporters are configured per Kafka client or broker. This makes it hard to use reporters in client-side applications that use multiple clients such as Connect, MirrorMaker, Streams but also Strimzi components like the HTTP bridge. At the moment, for these components it's preferable to keep the current metrics pipeline with JMX and jmx_exporter. For these two reasons, I propose only using reporters for brokers/controllers and only using them in KRaft mode.
+
+## Affected/not affected projects
+
+This affects `strimzi-kafka-operator `. Reporters are not usable with ZooKeeper and currently don't work well with client side application like Kafka Connect, MirrorMaker and the HTTP bridge. So Strimzi should keep support for jmx_exporter. We may reconsider dropping support once ZooKeeper is not supported anymore and if improvements in Apache Kafka make reporters usable with client side applications.
+
+The reporter will be a new project and should be usable without Strimzi. As it's new code, it should be first release as a preview feature (disabled by default). Once KRaft becomes the preferred deployment option, we should make the reporter the recommended option to expose metrics for brokers and controllers.
+
+In `strimzi-kafka-operator`, we will need new examples demonstrating how to use the reporter, new grafana dashboards with metric names from the reporter.
+
+Due to the way the OAuth plugin exports its metrics, the Prometheus reporter won't be usable with the plugin. It will require [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics) to work.
```

Comment:
What is the plan for this KIP? Can you add some rough timeline? While there is still some time before KRaft is production ready, I think this is a blocker for using this in KRaft by default from the start.

I assume the same applies for https://github.com/StyraInc/opa-kafka-plugin which is the other plugin we ship out of the box with metrics support.